From: Richard Acayan <mailingradian@gmail.com>
Date: Wed, 30 Jul 2025 20:08:32 -0400
Subject: [PATCH] media: venus: use a null pointer on HFI 4xx

On the Pixel 3a (SDM670), the Venus firmware tries to handle buffers
under Content Protection (CP) differently from other regular buffers,
according to each buffer's physical address. This doesn't work with
the hardcoded EOS buffer, 0xDEADB000. Use the null address (0) for
HFI 4xx.

A different implementation allocates a 4K buffer for sending an EOS, but
it is more complex than using a different pointer that works.

Sample error dmesg:

        [ 1889.304675] qcom-venus aa00000.video-codec: session error: event id:1001 (deadb000), session id:f6c9ac0d
        [ 1889.304699] qcom-venus-decoder aa00000.video-codec:video-core0: dec: event session error 1001

Error message from Venus firmware (HFI_MSG_SYS_DEBUG message):

        <VFW_E:HostDr:264d:f6c9ac0d:0> IsBufferProtected(1087): VenusCoreCtrl: CP_UNKNOWN, buffer straddles CP & non-CP regions

Signed-off-by: Richard Acayan <mailingradian@gmail.com>
---
 drivers/media/platform/qcom/venus/vdec.c | 3 ++-
 drivers/media/platform/qcom/venus/venc.c | 5 ++++-
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/media/platform/qcom/venus/vdec.c b/drivers/media/platform/qcom/venus/vdec.c
index 27cb024..ae4fba0 100644
--- a/drivers/media/platform/qcom/venus/vdec.c
+++ b/drivers/media/platform/qcom/venus/vdec.c
@@ -573,7 +573,8 @@ vdec_decoder_cmd(struct file *file, void *fh, struct v4l2_decoder_cmd *cmd)
 		 * avoids misapplying this quirk and breaking VP9 decode on SC7280.
 		 */
 
-		if (IS_IRIS2(inst->core) && is_fw_rev_or_older(inst->core, 1, 0, 87))
+		if ((IS_IRIS2(inst->core) && is_fw_rev_or_older(inst->core, 1, 0, 87))
+		 || IS_V4(inst->core))
 			fdata.device_addr = 0;
 		else
 			fdata.device_addr = 0xdeadb000;
diff --git a/drivers/media/platform/qcom/venus/venc.c b/drivers/media/platform/qcom/venus/venc.c
index fba0755..b415671 100644
--- a/drivers/media/platform/qcom/venus/venc.c
+++ b/drivers/media/platform/qcom/venus/venc.c
@@ -547,7 +547,10 @@ venc_encoder_cmd(struct file *file, void *fh, struct v4l2_encoder_cmd *cmd)
 
 		fdata.buffer_type = HFI_BUFFER_INPUT;
 		fdata.flags |= HFI_BUFFERFLAG_EOS;
-		fdata.device_addr = 0xdeadb000;
+		if (IS_V4(inst->core))
+			fdata.device_addr = 0;
+		else
+			fdata.device_addr = 0xdeadb000;
 
 		ret = hfi_session_process_buf(inst, &fdata);
 
